ledger/sandbox-on-x/src/main/scala/com/daml/ledger/sandbox/ConflictCheckingLedgerBridge.scala

```diff
@@ -0,0 +1,595 @@
+  ): Future[Either[SoxRejection, Unit]] = {
+    val informees = transaction.transaction.informees
+    indexService
+      .getParties(informees.toSeq)
+      .map { partyDetails =>
+        val allocatedInformees = partyDetails.iterator.map(_.party).toSet
+        if (allocatedInformees == informees) Right(())
+        else
+          Left(
+            SoxRejection.UnallocatedParties((informees diff allocatedInformees).toSet)(transaction)
+          )
+      }
+  }
+
+  private def checkTimeModel(
+      transaction: SubmittedTransaction
+  ): Future[Either[SoxRejection, Unit]] = {
+    val _ = transaction
+    // TODO implement
+    Future.successful(Right(()))
+  }
+
+  private def validateCausalMonotonicity(
+      transaction: Transaction,
+      inputContracts: Set[ContractId],
+      transactionLedgerEffectiveTime: Timestamp,
+      divulged: Set[ContractId],
+  ): Future[Either[SoxRejection, Unit]] = {
+    val referredContracts = inputContracts.diff(divulged)
+    if (referredContracts.isEmpty)
+      Future.successful(Right(()))
+    else
+      indexService
+        .lookupMaximumLedgerTime(referredContracts)
+        .transform {
+          case Failure(MissingContracts(missingContractIds)) =>
+            Success(Left(UnknownContracts(missingContractIds)(transaction)))
+          case Failure(err) =>
+            Success(Left(GenericRejectionFailure(err.getMessage)(transaction)))
+          case Success(maximumLedgerEffectiveTime) =>
+            maximumLedgerEffectiveTime
+              .filter(_ > transactionLedgerEffectiveTime)
+              .fold[Try[Either[SoxRejection, Unit]]](Success(Right(())))(
+                contractLedgerEffectiveTime =>
+                  Success(
+                    Left(
+                      CausalMonotonicityViolation(
+                        contractLedgerEffectiveTime = contractLedgerEffectiveTime,
+                        transactionLedgerEffectiveTime = transactionLedgerEffectiveTime,
+                      )(transaction)
+                    )
+                  )
+              )
+        }
+  }
+
+  private def validateKeyUsages(
+      transaction: Transaction,
+      keyInputs: KeyInputs,
+  ): Future[Either[SoxRejection, Unit]] =
+    keyInputs
+      .map(_.foldLeft(Future.successful[Either[SoxRejection, Unit]](Right(()))) {
+        case (f, (key, inputState)) =>
+          f.flatMap {
+            case Right(_) => validateExpectedKey(transaction, key, inputState)
+            case left => Future.successful(left)
+          }
+      })
+      .fold(
+        err => Future.successful(Left(GenericRejectionFailure(err.toString)(transaction))),
+        identity,
+      )
+
+  private def validateExpectedKey(
+      transaction: Submission.Transaction,
+      key: Key,
+      inputState: TxKeyInput,
+  ) =
+    indexService
+      // TODO SoX: Is it fine to use informees as readers?
+      .lookupContractKey(transaction.transaction.informees, key)
```

Comment:
It should semantically work, but might be a bit expensive, in particular for transactions with many informees.

The better option would be to use a contract-key lookup that is not guarded by the readers.